java/client/src/test/java/glide/api/RedisClusterClientTest.java

```diff
@@ -149,6 +153,61 @@ public <T> CompletableFuture<T> submitCommandToChannel(
         }
     }
 
+    @SneakyThrows
+    @Test
+    public void exec_without_routing() {
+        // setup
+        Object[] value = new Object[] {"PONG", "PONG"};
+        ClusterTransaction transaction = new ClusterTransaction().ping().ping();
+
+        CompletableFuture<Object[]> testResponse = new CompletableFuture<>();
+        testResponse.complete(value);
+
+        // match on protobuf request
+        when(commandManager.<Object[]>submitNewCommand(eq(transaction), eq(Optional.empty()), any()))
+                .thenReturn(testResponse);
+
+        // exercise
+        CompletableFuture<Object[]> response = service.exec(transaction);
+        Object[] payload = response.get();
+
+        // verify
+        assertEquals(testResponse, response);
+        assertArrayEquals(value, payload);
+    }
+
+    @SneakyThrows
+    @Test
+    public void exec_with_routing() {
+        // setup
+        Object[] value = new Object[] {"PONG", "PONG"};
+        ClusterTransaction transaction = new ClusterTransaction().ping().ping();
+        Route route = RANDOM;
+
+        CompletableFuture<Object[]> testResponse = new CompletableFuture<>();
+        testResponse.complete(value);
+
+        // match on protobuf request
+        when(commandManager.<Object[]>submitNewCommand(eq(transaction), eq(Optional.of(route)), any()))
+                .thenReturn(testResponse);
+
+        // exercise
+        CompletableFuture<Object[]> response = service.exec(transaction, route);
+        Object[] payload = response.get();
+
+        // verify
+        assertEquals(testResponse, response);
+        assertArrayEquals(value, payload);
+    }
+
+    @SneakyThrows
+    @Test
+    public void exec_supports_only_single_node_route() {
+        var exception =
```

Comment:
What is the exception type here do we need it to be uniform across clients?

Reply:
It is dummy runtime exception. Pyhton and node clients have type checking there, so you can't call `exec` with wrong routing type

Reply:
Fixed in the next commit